### PR TITLE
[REEF-1787] Unmanaged REEF on Spark packages global.jar resources incorrectly

### PR DESCRIPTION
**Pull request by motus**
Fix `JARFileMaker` class to handle symlinks to symlinks properly; add logging and refactor for readability and maintainability

**NOTE:** This fix allows REEF-on-Spark to run correctly.

JIRA: [REEF-1787](https://issues.apache.org/jira/browse/REEF-1787)

### CODE REVIEW

File: lang/java/reef-common/src/main/java/org/apache/reef/util/JARFileMaker.java [hunk 1]
```diff
@@ -34,87 +34,83 @@
...
 
-  private String makeRelative(final File input) throws IOException {
-    final String result;
-    if (this.relativeStartCanonicalPath == null) {
-      result = input.getCanonicalPath();
-    } else {
-      result = input.getCanonicalPath()
-          .replace(this.relativeStartCanonicalPath, "") // Drop the absolute prefix
-          .substring(1);                                // drop the '/' at the beginning
+  private static String internalJarPath(final File inputFile, final String prefix) {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/util/JARFileMaker.java]: Please make this method name a verb, e.g. `createPathInJar()` or such,

File: lang/java/reef-common/src/main/java/org/apache/reef/util/JARFileMaker.java [hunk 2]
```diff
@@ -34,87 +34,83 @@
...
-      final File[] files = inputFile.listFiles();
-      if (files != null) {
-        for (final File nestedFile : files) {
-          add(nestedFile);
-        }
+  public JARFileMaker addChildren(final File folder) throws IOException {
+    LOG.log(Level.FINEST, "Add children: {0}", folder);
+    final File[] files = folder.listFiles();
+    if (files != null) {
+      for (final File nestedFile : files) {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/util/JARFileMaker.java]: We could make this slightly nicer using `commons-lang` `ArrayUtils`:

```java
for (final File nestedFile :  ArrayUtils.nullToEmpty(files)) {
...
}
```

This way, we don't need the `if` anymore.

File: lang/java/reef-common/src/main/java/org/apache/reef/util/JARFileMaker.java [hunk 3]
```diff
@@ -34,87 +34,83 @@
...
-  public JARFileMaker addChildren(final File folder) throws IOException {
-    this.relativeStartCanonicalPath = folder.getCanonicalPath();
-    final File[] files = folder.listFiles();
-    if (files != null) {
-      for (final File f : files) {
-        this.add(f);
+    if (inputFile.isDirectory()) {
+      this.jarOutputStream.closeEntry();
+      final File[] files = inputFile.listFiles();
+      if (files != null) {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/util/JARFileMaker.java]: Same as above. Consider rewriting as:

```java
for (final File nestedFile :  ArrayUtils.nullToEmpty(files)) {
...
}
```

File: lang/java/reef-common/src/main/java/org/apache/reef/util/JARFileMaker.java (hunk 2 above)
Comment (motus) [inline lang/java/reef-common/src/main/java/org/apache/reef/util/JARFileMaker.java]: Hmm, it actually doesn't work without a cast:
```java
for (final File nestedFile :  (File[])ArrayUtils.nullToEmpty(files)) { // ...
```
Are you sure we want it like that?

**Comment by motus** [comment]
OK, I think I am done with the fixes - @markusweimer please take a look.

**Comment by markusweimer** [comment]
I'll merge this.

**Comment by markusweimer** [comment]
I merged this, but JIRA is down.

**Pull request closed without merging**